File: app/src/main/java/com/supercilex/robotscouter/model/scout/viewholders/EditTextViewHolder.java
package com.supercilex.robotscouter.model.scout.viewholders;

import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.google.firebase.database.DatabaseReference;
import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.model.scout.metrics.ScoutEditText;
import com.supercilex.robotscouter.model.scout.metrics.ScoutMetric;

public class EditTextViewHolder extends ScoutViewHolder {
    private TextView mName;
    private EditText mNotes;

    public EditTextViewHolder(View itemView) {
        super(itemView);
        mName = (TextView) itemView.findViewById(R.id.current_scout_text_view);
        mNotes = (EditText) itemView.findViewById(R.id.current_scout_edit_text);
    }

    @Override
    public void initialize(ScoutMetric view, DatabaseReference ref) {
        ScoutEditText scoutEditText = (ScoutEditText) view;

        setText(scoutEditText.getName());
        setValue(scoutEditText.getValue());
        setOnFocusChangeListener(scoutEditText, ref);
    }

    public void setText(String name) {
        mName.setText(name);
    }

    private void setValue(String value) {
        mNotes.setText(value);
    }

    private void setOnFocusChangeListener(final ScoutEditText scoutEditText,
                                          final DatabaseReference ref) {
        mNotes.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (!hasFocus) {
                    scoutEditText.setValue(ref, mNotes.getText().toString());
                }
            }
        });
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scouting/scout/viewholder/EditTextViewHolder.kt
package com.supercilex.robotscouter.ui.scouting.scout.viewholder

import android.os.Build
import android.support.annotation.CallSuper
import android.support.design.widget.TextInputLayout
import android.view.View
import android.widget.TextView
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.Metric
import com.supercilex.robotscouter.ui.scouting.MetricViewHolderBase

open class EditTextViewHolder(itemView: View) :
        MetricViewHolderBase<Metric.Text, String?, TextView>(itemView),
        View.OnFocusChangeListener {
    private val textLayout: TextInputLayout = itemView.findViewById(R.id.text_layout)

    public override fun bind() {
        super.bind()
        name.text = metric.value
        textLayout.hint = metric.name
        name.onFocusChangeListener = this

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                && metric.name.toUpperCase().contains(View.AUTOFILL_HINT_NAME.toUpperCase())) {
            itemView.importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_AUTO
            name.importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_YES
            name.setAutofillHints(View.AUTOFILL_HINT_NAME)
        }
    }

    @CallSuper
    override fun onFocusChange(v: View, hasFocus: Boolean) {
        if (!hasFocus) updateMetricValue(name.text.toString())
    }
}
